README.md
# Edfize

[![Build Status](https://travis-ci.com/sleepepi/edfize.svg?branch=master)](https://travis-ci.com/sleepepi/edfize)
[![Code Climate](https://codeclimate.com/github/sleepepi/edfize/badges/gpa.svg)](https://codeclimate.com/github/sleepepi/edfize)

Ruby gem used to load, validate, and parse European Data Format files. Used for
batch testing EDFs for errors. Ruby 2.6+ compatible.

## Installation

Use `gem install edfize` to update Edfize to the latest stable

Use `gem install edfize --pre` to update Edfize to the latest prerelease

## Usage

### Validate EDFs

Use `edfize test` to test that EDFs stored in the current directory have a valid
format.

    cd <edf-directory>
    edfize test

A list of validations performed is:

- **Expected Length Check**: Compares the calculated size of the file based on
  signal sizes defined in the header with the actual file size. A failure may
  indicate corruption in the header (if the expected is less than the actual
  file size), or a partial/truncated file (if the expected is more than the
  actual file size).
- **Reserved Area Checks**: Check that reserved areas are blank. Non-blank
  reserved areas can indicate a sign of header or EDF file corruption.

Flags that can be added to the `test` command include:

- `--failing`: Only display EDFs with failing tests
- `--quiet`: Suppress detailed failure descriptions that show the expected
  versus the actual result of the test

### Print Signal Header information

Use `edfize run` to print out signal header information for each EDF in the
current directory.

    cd <edf-directory>
    edfize run

### View A List of All Available Commands for Edfize

Use `edfize help` to list available commands and descriptions for Edfize.

    edfize help

### View Current Version of Edfize

Use `edfize version` to check the version of Edfize.

    edfize version

### Example of how to rewrite the start date of recording for a folder of EDFs

This will update all EDFs in the current directory and subdirectories with a
start date of 1st Jan 1985.

`rewrite_signal_date.rb`
```ruby
# gem install edfize --no-document
# ruby rewrite_signal_date.rb

require "rubygems"
require "edfize"

CLIPPING_DATE = "01.01.85"

Edfize.edfs do |edf|
  edf.update(start_date_of_recording: CLIPPING_DATE)
end
```

### Example of how to Load and Analyze EDFs in a Ruby Script

The following Ruby file demonstrates how to make use of the Edfize gem to load
EDF signals into arrays for analysis.

`tutorial_01_load_edf_and_signals.rb`
```ruby
# Tutorial 01 - Load EDF and Signals
#
#   gem install edfize --no-document
#
#   ruby tutorial_01_load_edf_and_signals.rb
#
# The EDF exists at:
#
#    https://sleepdata.org/datasets/shhs/files/edfs/shhs1?f=shhs1-200001.edf
#

require "rubygems"
require "edfize"

# Loads the file and reads the EDF Header
edf = Edfize::Edf.new("shhs1-200001.edf")

# Loads the data section of the EDF into Signal objects
edf.load_signals

# Print out information on the signals
puts "EDF #{edf.filename} contains the following #{edf.signals.count} signal#{"s" unless edf.signals.count == 1}:\n\n"

edf.signals.each do |signal|
  puts "Signal"
  puts "  Label                    : #{signal.label}"
  puts "  Samples Per Data Record  : #{signal.samples_per_data_record}"
  puts "  First 10 Physical Values : #{(signal.physical_values[0..10] + ["..."]).inspect}\n\n"
end
```

When run, the code above will output the following:

```console
EDF shhs1-200001.edf contains the following 14 signals:

Signal
  Label                    : SaO2
  Samples Per Data Record  : 1
  First 10 Physical Values : [95.31242847333486, 95.31242847333486, 95.31242847333486, 95.31242847333486, 95.31242847333486, 95.31242847333486, 95.31242847333486, 95.31242847333486, 94.14053559166858, 94.14053559166858, 94.14053559166858, "..."]

Signal
  Label                    : H.R.
  Samples Per Data Record  : 1
  First 10 Physical Values : [77.34416723887999, 77.34416723887999, 77.34416723887999, 76.56595712214848, 76.56595712214848, 76.56595712214848, 75.00190737773708, 75.00190737773708, 75.00190737773708, 75.00190737773708, 75.00190737773708, "..."]

Signal
  Label                    : EEG(sec)
  Samples Per Data Record  : 125
  First 10 Physical Values : [-4.411764705882348, 5.392156862745111, 2.4509803921568647, 0.49019607843136725, -0.49019607843136725, -10.294117647058826, 3.4313725490196134, 12.25490196078431, -1.470588235294116, -2.4509803921568647, -8.333333333333329, "..."]

Signal
  Label                    : ECG
  Samples Per Data Record  : 125
  First 10 Physical Values : [0.03431372549019618, 0.03431372549019618, 0.03431372549019618, 0.03431372549019618, 0.044117647058823595, 0.044117647058823595, 0.044117647058823595, 0.044117647058823595, 0.044117647058823595, 0.03431372549019618, 0.03431372549019618, "..."]

Signal
  Label                    : EMG
  Samples Per Data Record  : 125
  First 10 Physical Values : [12.622549019607845, 3.7990196078431353, -3.5539215686274517, -2.5735294117647065, 8.455882352941174, 1.5931372549019613, 9.436274509803923, -8.700980392156861, -2.5735294117647065, 13.112745098039213, -12.867647058823529, "..."]

Signal
  Label                    : EOG(L)
  Samples Per Data Record  : 50
  First 10 Physical Values : [28.921568627450966, 17.15686274509804, 25.0, 19.117647058823536, -5.392156862745097, -9.313725490196077, -0.49019607843136725, -1.470588235294116, 1.470588235294116, -1.470588235294116, 0.49019607843136725, "..."]

Signal
  Label                    : EOG(R)
  Samples Per Data Record  : 50
  First 10 Physical Values : [12.25490196078431, 1.470588235294116, 10.294117647058812, 5.392156862745111, 17.15686274509804, 18.137254901960773, 25.980392156862735, 32.84313725490196, 25.0, 26.960784313725497, 22.058823529411768, "..."]

Signal
  Label                    : EEG
  Samples Per Data Record  : 125
  First 10 Physical Values : [-2.4509803921568647, 1.470588235294116, -9.313725490196077, -6.372549019607845, -0.49019607843136725, -10.294117647058826, -12.25490196078431, -12.25490196078431, -7.352941176470594, 1.470588235294116, 6.372549019607845, "..."]

Signal
  Label                    : THOR RES
  Samples Per Data Record  : 10
  First 10 Physical Values : [0.207843137254902, 0.207843137254902, 0.15294117647058825, 0.0980392156862745, 0.03529411764705881, -0.0117647058823529, -0.050980392156862786, -0.08235294117647052, -0.10588235294117654, -0.1215686274509804, -0.13725490196078427, "..."]

Signal
  Label                    : ABDO RES
  Samples Per Data Record  : 10
  First 10 Physical Values : [0.30980392156862746, 0.24705882352941178, 0.16078431372549018, 0.06666666666666665, -0.0039215686274509665, -0.08235294117647052, -0.1607843137254903, -0.2078431372549019, -0.2313725490196079, -0.2549019607843137, -0.2705882352941176, "..."]

Signal
  Label                    : POSITION
  Samples Per Data Record  : 1
  First 10 Physical Values : [2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, "..."]

Signal
  Label                    : LIGHT
  Samples Per Data Record  : 1
  First 10 Physical Values : [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, "..."]

Signal
  Label                    : NEW AIR
  Samples Per Data Record  : 10
  First 10 Physical Values : [6.372549019607845, 6.372549019607845, 5.392156862745111, 3.4313725490196134, 7.35294117647058, 6.372549019607845, 8.333333333333343, 9.313725490196077, 6.372549019607845, 6.372549019607845, 7.35294117647058, "..."]

Signal
  Label                    : OX stat
  Samples Per Data Record  : 1
  First 10 Physical Values : [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, "..."]
```

## Contributing

1. Fork it ( https://github.com/sleepepi/edfize/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
## 0.6.0 (March 1, 2019)

### Enhancements
- **Gem Changes**
  - Update to ruby 2.6.1
  - Update to simplecov 0.16.1
  - Remove colorize

## 0.5.0 (April 21, 2017)

### Enhancements
- **Scripting Changes**
  - Removed `edfs_in_current_directory_and_subdirectories` in favor of
    `edf_paths`
- **Test Changes**
  - Running `edfize test --failing` now displays progress, by @kitlawes
- **Gem Changes**
  - Updated to Ruby 2.4.1
  - Updated to simplecov 0.14.1

## 0.4.0 (December 8, 2016)

### Enhancements
- **Scripting Changes**
  - Edfize now provides an enumerator to iterate through a folder of EDFs:
    - `Edfize.edfs` # Current directory and subdirectories
    - `Edfize.edfs(recursive: false)` # Only current directory
  - Deprecated `edfs_in_current_directory_and_subdirectories` in favor of
    `edf_paths`
    - Note that while `edf_paths` is a direct replacement for
      `edfs_in_current_directory_and_subdirectories`, it is recommended to use
      the new `edfs` enumerator instead
  - Added `Edf::update` method that allows portions of the header to be updated
    - Change an EDF's start_date_of_recording to the clipping date, 1 Jan 1985
    - `edf.update(start_date_of_recording: '01.01.85')`
    - Change an EDF's start_time_of_recording to midnight
    - `edf.update(start_time_of_recording: '00.00.00')`
- **Test Changes**
  - Removed `reserved_signal_areas_blank` check as it was flagging too many
    false positives

## 0.3.0 (December 7, 2016)

### Enhancements
- **Test Changes**
  - Added a test to make sure date in EDF and EDF+ formats are valid
    - The EDF header date is checked in the following manner: (a) Is the EDF+
      header (29-FEB-2100) a valid date, if (yes) check that the EDF header date
      (29.02.00) matches the EDF+ date, if (no) check that the EDF header date
      (29.02.00) is a valid date. The EDF+ header date is checked first since
      (29-FEB-2100) is NOT a valid date as leap years only occur in a century
      years divisible by 400, while the truncated EDF date would be (02.29.00)
      which would be incorrectly seen as valid since 2000 is a leap year.
- **Gem Changes**
  - Updated to Ruby 2.3.3
  - Updated to colorize 0.8.1
  - Updated simplecov to 0.11.2
  - Removed minitest-reporters
  - Updated to simplecov 0.12.0
  - Updated to bundler 1.13

### Bug Fix
- Fixed an issue where EDFs using `.EDF` instead of `.edf` weren't being tested
  by `edfize t`

## 0.2.0 (March 12, 2015)

### Enhancements
- EDFs can now be partially loaded by specifying the epoch number and epoch size
  in seconds
- Use of Ruby 2.2.1 is now recommended

### Testing
- Updated to minitest test framework along with simplecov for coverage

## 0.1.0 (May 28, 2014)
- Initial EDF class to load headers and signals into Ruby objects
- `edfize` command has the following actions:
  - `test`: Validates EDFs in the current directory and subdirectories for
    errors
    - To only show failing tests, add the flag `--failing`
      - `edfize test --failing`
    - To suppress descriptive test failures, add the flag `--quiet`
      - `edfize test --quiet`
    - Both flags can be used together
      - `edfize test --failing --quiet`
  - `run`: Prints out the headers of all EDFs in the current directory and
    subdirectories
  - `help`: Displays information about `edfize` along with all available
    commands
  - `version`: Displays the current `edfize` version
- `edfize test` checks for the following:
  - Expected Length Check: The expected total size is computed from the (`number
    of data records` * `total samples across all signals`) + `size of header`
    and this is compared to the actual file size.
  - Reserved Area Checks: The header and the individual header reserved areas
    are checked to validate that they are blank. Non-blank areas are a sign that
    the EDF header is corrupt and that data from the signal data block have
    leaked into the header itself.


test/edf_test.rb
# frozen_string_literal: true

require "test_helper"
require "tempfile"

# Test to assure EDFs can be loaded and updated.
class EdfTest < Minitest::Test
  def setup
    @valid_edf_no_data_records = Edfize::Edf.new("test/support/zero-data-records.edf")
    @valid_edf_with_three_signals = Edfize::Edf.new("test/support/simulated-01.edf")
    @edf_invalid_date = Edfize::Edf.new("test/support/invalid-date.edf")
  end

  def test_edf_version
    assert_equal 0, @valid_edf_no_data_records.send("compute_offset", :version)
    assert_equal 0, @valid_edf_no_data_records.version
  end

  def test_edf_local_patient_identification
    assert_equal 8, @valid_edf_no_data_records.send("compute_offset", :local_patient_identification)
    assert_equal "", @valid_edf_no_data_records.local_patient_identification
  end

  def test_edf_local_recording_identification
    assert_equal 88, @valid_edf_no_data_records.send("compute_offset", :local_recording_identification)
    assert_equal "", @valid_edf_no_data_records.local_recording_identification
  end

  def test_edf_start_date_of_recording
    assert_equal 168, @valid_edf_no_data_records.send("compute_offset", :start_date_of_recording)
    assert_equal "31.01.85", @valid_edf_no_data_records.start_date_of_recording
  end

  def test_edf_start_date
    assert_equal Date.parse("1985-01-31"), @valid_edf_no_data_records.start_date
    assert_nil @edf_invalid_date.start_date
  end

  def test_edf_start_time_of_recording
    assert_equal 176, @valid_edf_no_data_records.send("compute_offset", :start_time_of_recording)
    assert_equal "20.14.57", @valid_edf_no_data_records.start_time_of_recording
  end

  def test_edf_number_of_bytes_in_header
    assert_equal 184, @valid_edf_no_data_records.send("compute_offset", :number_of_bytes_in_header)
    assert_equal 3840, @valid_edf_no_data_records.number_of_bytes_in_header
  end

  def test_edf_reserved
    assert_equal 192, @valid_edf_no_data_records.send("compute_offset", :reserved)
    assert_equal " " * 44, @valid_edf_no_data_records.reserved
  end

  def test_edf_number_of_data_records
    assert_equal 236, @valid_edf_no_data_records.send("compute_offset", :number_of_data_records)
    assert_equal 0, @valid_edf_no_data_records.number_of_data_records
  end

  def test_edf_duration_of_a_data_record
    assert_equal 244, @valid_edf_no_data_records.send("compute_offset", :duration_of_a_data_record)
    assert_equal 1, @valid_edf_no_data_records.duration_of_a_data_record
  end

  def test_edf_number_of_signals
    assert_equal 252, @valid_edf_no_data_records.send("compute_offset", :number_of_signals)
    assert_equal 14, @valid_edf_no_data_records.number_of_signals
  end

  # Signal Header Tests

  def test_edf_signal_labels
    assert_equal 0, @valid_edf_no_data_records.send("compute_signal_offset", :label)
    assert_equal [
      "SaO2", "H.R.", "EEG(sec)", "ECG", "EMG", "EOG(L)", "EOG(R)", "EEG",
      "THOR RES", "ABDO RES", "POSITION", "LIGHT", "NEW AIR", "OX stat"
    ], @valid_edf_no_data_records.signals.collect(&:label)
  end

  def test_edf_signal_transducer_types
    assert_equal 16, @valid_edf_no_data_records.send("compute_signal_offset", :transducer_type)
    assert_equal [""] * 14, @valid_edf_no_data_records.signals.collect(&:transducer_type)
  end

  def test_edf_signal_physical_dimensions
    assert_equal 96, @valid_edf_no_data_records.send("compute_signal_offset", :physical_dimension)
    assert_equal [
      "", "", "uV", "mV", "uV", "uV", "uV", "uV", "", "", "", "", "uV", ""
    ], @valid_edf_no_data_records.signals.collect(&:physical_dimension)
  end

  def test_edf_signal_physical_minimums
    assert_equal 104, @valid_edf_no_data_records.send("compute_signal_offset", :physical_minimum)
    assert_equal [
      0.0, 0.0, -125.0, -1.25, -31.25, -125.0, -125.0, -125.0, 1.0, 1.0, 0.0, 0.0, -125.0, 0.0
    ], @valid_edf_no_data_records.signals.collect(&:physical_minimum)
  end

  def test_edf_signal_physical_maximums
    assert_equal 112, @valid_edf_no_data_records.send("compute_signal_offset", :physical_maximum)
    assert_equal [
      100.0, 250.0, 125.0, 1.25, 31.25, 125.0, 125.0, 125.0, -1.0, -1.0, 3.0, 1.0, 125.0, 3.0
    ], @valid_edf_no_data_records.signals.collect(&:physical_maximum)
  end

  def test_edf_signal_digital_minimums
    assert_equal 120, @valid_edf_no_data_records.send("compute_signal_offset", :digital_minimum)
    assert_equal [
      -32_768, -32_768, -128, -128, -128, -128, -128, -128, -128, -128, 0, 0, -128, 0
    ], @valid_edf_no_data_records.signals.collect(&:digital_minimum)
  end

  def test_edf_signal_digital_maximums
    assert_equal 128, @valid_edf_no_data_records.send("compute_signal_offset", :digital_maximum)
    assert_equal [
      32_767, 32_767, 127, 127, 127, 127, 127, 127, 127, 127, 3, 1, 127, 3
    ], @valid_edf_no_data_records.signals.collect(&:digital_maximum)
  end

  def test_edf_signal_prefilterings
    assert_equal 136, @valid_edf_no_data_records.send("compute_signal_offset", :prefiltering)
    assert_equal [""] * 14, @valid_edf_no_data_records.signals.collect(&:prefiltering)
  end

  def test_edf_signal_samples_per_data_records
    assert_equal 216, @valid_edf_no_data_records.send("compute_signal_offset", :samples_per_data_record)
    assert_equal [
      1, 1, 125, 125, 125, 50, 50, 125, 10, 10, 1, 1, 10, 1
    ], @valid_edf_no_data_records.signals.collect(&:samples_per_data_record)
  end

  def test_edf_signal_reserved_areas
    assert_equal 224, @valid_edf_no_data_records.send("compute_signal_offset", :reserved_area)
    assert_equal [" " * 32] * 14, @valid_edf_no_data_records.signals.collect(&:reserved_area)
  end

  def test_loads_single_epoch
    # Load the first epoch (0 index), with the epoch size being 1 second
    @valid_edf_with_three_signals.load_epoch(0, 1)
    @signal_one   = @valid_edf_with_three_signals.signals[0]
    @signal_two   = @valid_edf_with_three_signals.signals[1]
    @signal_three = @valid_edf_with_three_signals.signals[2]
    assert_equal [0, 1], @signal_one.digital_values
    assert_equal [50.000762951094835, 50.002288853284504], @signal_one.physical_values
    assert_equal [0, 1], @signal_two.digital_values
    assert_equal [125.00190737773708, 125.00572213321126], @signal_two.physical_values
    assert_equal [0, 1, 2, 3], @signal_three.digital_values
    assert_equal [
      0.49019607843136725, 1.470588235294116, 2.4509803921568647, 3.4313725490196134
    ], @signal_three.physical_values
  end

  def test_loads_last_epoch
    # Load the last epoch (0 index), with the epoch size being 1 second
    @valid_edf_with_three_signals.load_epoch(9, 1)
    @signal_one   = @valid_edf_with_three_signals.signals[0]
    @signal_two   = @valid_edf_with_three_signals.signals[1]
    @signal_three = @valid_edf_with_three_signals.signals[2]
    assert_equal [9, nil], @signal_one.digital_values
    assert_equal [50.01449607080186, nil], @signal_one.physical_values
    assert_equal [9, nil], @signal_two.digital_values
    assert_equal [125.03624017700466, nil], @signal_two.physical_values
    assert_equal [18, 19, nil, nil], @signal_three.digital_values
    assert_equal [18.137254901960773, 19.117647058823536, nil, nil], @signal_three.physical_values
  end

  def test_loads_last_epoch_of_two_seconds
    # Load the fourth epoch (0 index), with the epoch size being 2 seconds
    @valid_edf_with_three_signals.load_epoch(3, 2)
    @signal_one   = @valid_edf_with_three_signals.signals[0]
    @signal_two   = @valid_edf_with_three_signals.signals[1]
    @signal_three = @valid_edf_with_three_signals.signals[2]
    assert_equal [6, 7, 8], @signal_one.digital_values
    assert_equal [50.00991836423285, 50.01144426642252, 50.01297016861219], @signal_one.physical_values
    assert_equal [6, 7, 8], @signal_two.digital_values
    assert_equal [125.02479591058213, 125.02861066605631, 125.03242542153048], @signal_two.physical_values
    assert_equal [12, 13, 14, 15, 16, 17], @signal_three.digital_values
    assert_equal [
      12.25490196078431, 13.235294117647072, 14.215686274509807,
      15.196078431372541, 16.176470588235304, 17.15686274509804
    ], @signal_three.physical_values
  end

  def test_should_rewrite_start_date_of_recording
    file = Tempfile.new("invalid-date-copy.edf")
    FileUtils.cp("test/support/invalid-date.edf", file.path)
    edf = Edfize::Edf.new(file.path)
    assert_equal "00.00.00", edf.start_date_of_recording
    edf.update(start_date_of_recording: "01.01.85")
    edf_new = Edfize::Edf.new(file.path) # Load new EDF to check that change is written to disk.
    assert_equal "01.01.85", edf_new.start_date_of_recording
  ensure
    file.close
    file.unlink # Deletes temporary file.
  end

  def test_should_rewrite_start_time_of_recording
    file = Tempfile.new("invalid-date-copy.edf")
    FileUtils.cp("test/support/invalid-date.edf", file.path)
    edf = Edfize::Edf.new(file.path)
    assert_equal "20.14.57", edf.start_time_of_recording
    edf.update(start_time_of_recording: "12.34.56")
    edf_new = Edfize::Edf.new(file.path) # Load new EDF to check that change is written to disk.
    assert_equal "12.34.56", edf_new.start_time_of_recording
  ensure
    file.close
    file.unlink # Deletes temporary file.
  end
end


lib/edfize/tests.rb
# frozen_string_literal: true

require "edfize/color"
require "edfize/tests/result"
require "edfize/tests/runner"
require "edfize/tests/check_length"
require "edfize/tests/check_reserved_area"
require "edfize/tests/check_reserved_signal_areas"
require "edfize/tests/check_valid_date"

module Edfize
  module Tests
    extend CheckLength
    extend CheckReservedArea
    extend CheckReservedSignalAreas
    extend CheckValidDate
  end
end


test/edfize_test.rb
# frozen_string_literal: true

require "test_helper"
require "fileutils"

# Tests to assure that the gem commands run correctly.
class EdfizeTest < Minitest::Test
  def setup
    FileUtils.cd("test/support")
    @original_stdout = $stdout
    $stdout = StringIO.new
  end

  def teardown
    $stdout = @original_stdout
    @original_stdout = nil
    FileUtils.cd("../..")
  end

  def test_edfize_application
    assert_kind_of Module, Edfize
  end

  def test_edfize_version
    assert_kind_of String, Edfize::VERSION::STRING
  end

  def test_version_command
    assert_nil Edfize.launch(["v"])
  end

  def test_help_command
    assert_nil Edfize.launch(["h"])
  end

  def test_test_command
    assert_nil Edfize.launch(["t"])
  end

  def test_run_command
    assert_equal %w(invalid-date.edf simulated-01.edf zero-data-records.edf), Edfize.launch(["r"]).sort
  end
end


lib/edfize.rb
# frozen_string_literal: true

require "edfize/edf"
require "edfize/tests"
require "edfize/version"
require "edfize/color"

# Loads EDFs, prints information, runs tests
module Edfize
  def self.launch(argv)
    case argv.first.to_s.scan(/\w/).first
    when "v"
      version
    when "c", "t"
      check(argv[1..-1])
    when "r"
      print_headers
    else
      help
    end
  end

  def self.print_headers
    puts "----------------"
    edfs.each do |edf|
      edf.print_header
      puts "----------------"
    end
  end

  def self.version
    puts "Edfize #{Edfize::VERSION::STRING}"
  end

  def self.check(argv)
    test_start_time = Time.now
    edf_count = 0
    test_count = 0
    failure_count = 0
    total_edfs = edf_paths.count
    show_passing = argv.include?("--failing") ? false : true
    puts "Started\n"
    edfs.each do |edf|
      runner = Edfize::Tests::Runner.new(edf, argv)
      runner.run_tests
      test_count += runner.tests_run
      failure_count += runner.tests_failed
      edf_count += 1
      print "\rChecked EDF #{edf_count} of #{total_edfs}" unless show_passing || !runner.tests_failed.zero?
    end
    puts "\nFinished in #{Time.now - test_start_time}s"
    puts "#{edf_count} EDF#{"s" unless edf_count == 1}, #{test_count} test#{"s" unless test_count == 1}, " + "#{failure_count} failure#{"s" unless failure_count == 1}".send(failure_count == 0 ? :green : :red)
  end

  def self.help
    help_message = <<-EOT
Usage: edfize COMMAND [ARGS]

The most common edfize commands are:
  [t]est            Check EDFs in directory and subdirectories
    --failing       Only display failing tests
    --quiet         Suppress failing test descriptions
  [r]un             Print EDF header information
  [h]elp            Show edfize command documentation
  [v]ersion         Returns the version of Edfize

Commands can be referenced by the first letter:
  Ex: `edfize t`, for test

EOT
    puts help_message
  end

  # Returns an enumerator of EDFs.
  def self.edfs(recursive: true)
    return enum_for(:edfs, recursive: recursive) unless block_given?
    edf_paths(recursive: recursive).each do |file_path|
      yield Edf.new(file_path)
    end
  end

  def self.edf_paths(recursive: true)
    path = "#{"**/" if recursive}*.edf"
    Dir.glob(path, File::FNM_CASEFOLD)
  end
end


lib/edfize/edf.rb
# frozen_string_literal: true

require "edfize/signal"
require "date"

module Edfize
  # Class used to load and manipulate EDFs
  class Edf
    # EDF File Path
    attr_reader   :filename

    # Header Information
    attr_accessor :version
    attr_accessor :local_patient_identification
    attr_accessor :local_recording_identification
    attr_accessor :start_date_of_recording
    attr_accessor :start_time_of_recording
    attr_accessor :number_of_bytes_in_header
    attr_accessor :reserved
    attr_accessor :number_of_data_records
    attr_accessor :duration_of_a_data_record
    attr_accessor :number_of_signals

    attr_accessor :signals

    HEADER_CONFIG = {
      version:                        { size:  8, after_read: :to_i,  name: "Version" },
      local_patient_identification:   { size: 80, after_read: :strip, name: "Local Patient Identification" },
      local_recording_identification: { size: 80, after_read: :strip, name: "Local Recording Identification" },
      start_date_of_recording:        { size:  8,                     name: "Start Date of Recording", description: "(dd.mm.yy)" },
      start_time_of_recording:        { size:  8,                     name: "Start Time of Recording", description: "(hh.mm.ss)"},
      number_of_bytes_in_header:      { size:  8, after_read: :to_i,  name: "Number of Bytes in Header" },
      reserved:                       { size: 44,                     name: "Reserved" },
      number_of_data_records:         { size:  8, after_read: :to_i,  name: "Number of Data Records" },
      duration_of_a_data_record:      { size:  8, after_read: :to_i,  name: "Duration of a Data Record", units: "second" },
      number_of_signals:              { size:  4, after_read: :to_i,  name: "Number of Signals" }
    }

    HEADER_OFFSET = HEADER_CONFIG.collect{|k,h| h[:size]}.inject(:+)

    SIZE_OF_SAMPLE_IN_BYTES = 2

    # Used by tests
    RESERVED_SIZE = HEADER_CONFIG[:reserved][:size]

    def self.create(filename, &block)
      edf = new(filename)
      yield edf if block_given?
      edf
    end

    def initialize(filename)
      @filename = filename
      @signals = []

      read_header
      read_signal_header
      self
    end

    def load_signals
      get_data_records
    end

    # Epoch Number is Zero Indexed, and Epoch Size is in Seconds (Not Data Records)
    def load_epoch(epoch_number, epoch_size)
      # reset_signals!
      load_digital_signals_by_epoch(epoch_number, epoch_size)
      calculate_physical_values!
    end

    def size_of_header
      HEADER_OFFSET + ns * Signal::SIGNAL_CONFIG.collect { |_k, h| h[:size] }.inject(:+)
    end

    def expected_size_of_header
      @number_of_bytes_in_header
    end

    # Total File Size In Bytes
    def edf_size
      File.size(@filename)
    end

    # Data Section Size In Bytes
    def expected_data_size
      @signals.collect(&:samples_per_data_record).inject(:+).to_i * @number_of_data_records * SIZE_OF_SAMPLE_IN_BYTES
    end

    def expected_edf_size
      expected_data_size + size_of_header
    end

    def section_value_to_string(section)
      instance_variable_get("@#{section}").to_s
    end

    def section_units(section)
      units = HEADER_CONFIG[section][:units].to_s
      if units == ""
        ""
      else
        " #{units}" + (instance_variable_get("@#{section}") == 1 ? "" : "s")
      end
    end

    def section_description(section)
      description = HEADER_CONFIG[section][:description].to_s
      if description == ""
        ""
      else
        " #{description}"
      end
    end

    def print_header
      puts "\nEDF                            : #{@filename}"
      puts "Total File Size                : #{edf_size} bytes"
      puts "\nHeader Information"
      HEADER_CONFIG.each do |section, hash|
        puts "#{hash[:name]}#{" "*(31 - hash[:name].size)}: " + section_value_to_string(section) + section_units(section) + section_description(section)
      end
      puts "\nSignal Information"
      signals.each_with_index do |signal, index|
        puts "\n  Position                     : #{index + 1}"
        signal.print_header
      end
      puts "\nGeneral Information"
      puts "Size of Header (bytes)         : #{size_of_header}"
      puts "Size of Data   (bytes)         : #{data_size}"
      puts "Total Size     (bytes)         : #{edf_size}"

      puts "Expected Size of Header (bytes): #{expected_size_of_header}"
      puts "Expected Size of Data   (bytes): #{expected_data_size}"
      puts "Expected Total Size     (bytes): #{expected_edf_size}"
    end

    def start_date
      (dd, mm, yy) = start_date_of_recording.split(".")
      dd = parse_integer(dd)
      mm = parse_integer(mm)
      yy = parse_integer(yy)
      yyyy = if yy && yy >= 85
               yy + 1900
             else
               yy + 2000
             end
      Date.strptime("#{mm}/#{dd}/#{yyyy}", "%m/%d/%Y")
    rescue
      nil
    end

    def parse_integer(string)
      Integer(format("%g", string))
    rescue
      nil
    end

    def update(hash)
      hash.each do |section, value|
        update_header_section(section, value)
      end
    end

    def update_header_section(section, value)
      return false unless HEADER_CONFIG.keys.include?(section)
      send "#{section}=", value
      size = HEADER_CONFIG[section][:size]
      string = format("%-#{size}.#{size}s", send(section).to_s)
      IO.binwrite(filename, string, send(:compute_offset, section))
      true
    end

    protected

    def read_header
      HEADER_CONFIG.keys.each do |section|
        read_header_section(section)
      end
    end

    def read_header_section(section)
      result = IO.binread(@filename, HEADER_CONFIG[section][:size], compute_offset(section) )
      result = result.to_s.send(HEADER_CONFIG[section][:after_read]) unless HEADER_CONFIG[section][:after_read].to_s == ""
      self.instance_variable_set("@#{section}", result)
    end

    def compute_offset(section)
      offset = 0
      HEADER_CONFIG.each do |key, hash|
        break if key == section
        offset += hash[:size]
      end
      offset
    end

    def ns
      @number_of_signals
    end

    def reset_signals!
      @signals = []
      read_signal_header
    end

    def create_signals
      (0..ns-1).to_a.each do |signal_number|
        @signals[signal_number] ||= Signal.new()
      end
    end

    def read_signal_header
      create_signals
      Signal::SIGNAL_CONFIG.keys.each do |section|
        read_signal_header_section(section)
      end
    end

    def compute_signal_offset(section)
      offset = 0
      Signal::SIGNAL_CONFIG.each do |key, hash|
        break if key == section
        offset += hash[:size]
      end
      offset
    end

    def read_signal_header_section(section)
      offset = HEADER_OFFSET + ns * compute_signal_offset(section)
      (0..ns-1).to_a.each do |signal_number|
        section_size = Signal::SIGNAL_CONFIG[section][:size]
        result = IO.binread(@filename, section_size, offset+(signal_number*section_size))
        result = result.to_s.send(Signal::SIGNAL_CONFIG[section][:after_read]) unless Signal::SIGNAL_CONFIG[section][:after_read].to_s == ""
        @signals[signal_number].send("#{section}=", result)
      end
    end

    def get_data_records
      load_digital_signals()
      calculate_physical_values!()
    end

    def load_digital_signals_by_epoch(epoch_number, epoch_size)
      size_of_data_record_in_bytes = @signals.collect(&:samples_per_data_record).inject(:+).to_i * SIZE_OF_SAMPLE_IN_BYTES
      data_records_to_retrieve = (epoch_size / @duration_of_a_data_record rescue 0)
      length_of_bytes_to_read = (data_records_to_retrieve+1) * size_of_data_record_in_bytes
      epoch_offset_size = epoch_number * epoch_size * size_of_data_record_in_bytes # TODO: The size in bytes of an epoch

      all_signal_data = (IO.binread(@filename, length_of_bytes_to_read, size_of_header + epoch_offset_size).unpack("s<*") rescue [])
      load_signal_data(all_signal_data, data_records_to_retrieve+1)
    end

    # 16-bit signed integer size = 2 Bytes = 2 ASCII characters
    # 16-bit signed integer in "Little Endian" format (least significant byte first)
    # unpack:  s<         16-bit signed, (little-endian) byte order
    def load_digital_signals
      all_signal_data = IO.binread(@filename, nil, size_of_header).unpack("s<*")
      load_signal_data(all_signal_data, @number_of_data_records)
    end

    def load_signal_data(all_signal_data, data_records_retrieved)
      all_samples_per_data_record = @signals.collect{|s| s.samples_per_data_record}
      total_samples_per_data_record = all_samples_per_data_record.inject(:+).to_i

      offset = 0
      offsets = []
      all_samples_per_data_record.each do |samples_per_data_record|
        offsets << offset
        offset += samples_per_data_record
      end

      (0..data_records_retrieved-1).to_a.each do |data_record_index|
        @signals.each_with_index do |signal, signal_index|
          read_start = data_record_index * total_samples_per_data_record + offsets[signal_index]
          (0..signal.samples_per_data_record - 1).to_a.each do |value_index|
            signal.digital_values << all_signal_data[read_start+value_index]
          end
        end
      end
    end

    def calculate_physical_values!
      @signals.each{|signal| signal.calculate_physical_values!}
    end

    def data_size
      IO.binread(@filename, nil, size_of_header).size
    end
  end
end
